### PR TITLE
msg/async/AsyncConnection: unregister connection when racing happened

### PR DESCRIPTION
**Pull request by yuyuyu101**
Signed-off-by: Haomai Wang <haomai@xsky.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-11-21_16:17:37-rados-wip-kefu-testing-2017-11-21-1844-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-11-22_05:41:59-rados-wip-kefu-testing-2017-11-21-1844-distro-basic-mira/

**Comment by tchaikov** [comment]
@yuyuyu101 could you be more specific what kind of racing is expected here? if the existing conn is replaced, `existing->_stop()` will be called before `accept_conn()`, so it should have been unregistered from `this->conns` already.

**Comment by yuyuyu101** [comment]
no, it stop is called before accpet_conn, it will remove no elements in conns. then accept_conn will add conn later. so it will leak

**Comment by tchaikov** [comment]
> it will remove no elements in conns

but i think `async_msgr->unregister_conn(this);` does remove the conn from `deleted_conns`. and in that case, `deleted_conns.erase(existing)` returns `1`, so `conns.erase(it); will be executed. see https://github.com/ceph/ceph/pull/19013/files#diff-f77f790032acbbb064840b4efd71939dL357

**Comment by yuyuyu101** [comment]
thread 1: call _stop
thread 2: call accept_conn, then execute after "deleted_conns.erase" statement
thread 1: unregister_conn
thread 2: set element in conns

File: src/msg/async/AsyncMessenger.h [hunk 1]
```diff
@@ -356,10 +356,8 @@ class AsyncMessenger : public SimplePolicyMessenger {
       Mutex::Locker l(deleted_lock);
       if (deleted_conns.erase(existing)) {
         existing->get_perf_counter()->dec(l_msgr_active_connections);
-        conns.erase(it);
-      } else if (conn != existing) {
-        return -1;
       }
+      conns.erase(it);
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.h]: oic. so the `existing` is retrieved from `deleted_conn` and `conns`, and is not put back to `deleted_conn` yet. 

probably we can just remove the line above. `conns[conn->peer_addr] = conn;` will take care of this.

@yuyuyu101 what do you think?

**Comment by yuyuyu101** [comment]
yes, it's right. updated

**Comment by tchaikov** [comment]
@yuyuyu101 mind filing a ticket on tracker? so we can backport it to luminous.

**Comment by yuyuyu101** [comment]
http://tracker.ceph.com/issues/22437

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-12-14_08:53:42-rados-wip-kefu-testing-2017-12-14-1157-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-12-15_06:17:13-rados-wip-kefu-testing-2017-12-14-1157-distro-basic-mira/

**Pull request merged**